go/vt/vttablet/tabletmanager/vreplication/vreplicator.go

```diff
@@ -558,3 +570,298 @@ func recalculatePKColsInfoByColumnNames(uniqueKeyColumnNames []string, colInfos
+		}
+		id, err := res.Rows[0][0].ToInt64()
+		if err != nil {
+			return fmt.Errorf("unexpected connection ID value returned (%+v) when attempting to kill %q", res.Rows[0][0], alterStmt)
+		}
+		_, err = killdbc.ExecuteFetch(fmt.Sprintf("kill %d", id), 1)
+		return err
+	}
+	go func() {
+		select {
+		case <-ctx.Done():
+			log.Infof("Copy of %q table stopped when performing the following post copy action for the %q VReplication workflow: %+v",
+				tableName, vr.WorkflowName, action)
+			if err := killAlterStmt(action.Task); err != nil {
+				log.Errorf("Failed to kill post copy ALTER statement for the %q VReplication workflow: %v", vr.WorkflowName, err)
+			}
+			return
+		case <-done:
+			// We're done, so no longer need to listen for cancellation.
+			return
+		}
+	}()
+
+	for _, row := range qr.Named().Rows {
+		select {
+		case <-ctx.Done():
+			return vterrors.Errorf(vtrpcpb.Code_CANCELED, "context has expired")
+		default:
+		}
+		action = PostCopyAction{}
+		acv, err := row["action"].ToBytes()
+		if err != nil {
+			return err
+		}
+		if err := json.Unmarshal(acv, &action); err != nil {
+			return err
+		}
+
+		switch action.Type {
+		case PostCopyActionSQL:
+			log.Infof("Executing post copy SQL action for the %q table in the %q VReplication workflow: %s",
+				tableName, vr.WorkflowName, action.Task)
+			// This will return an io.EOF / MySQL CRServerLost (errno 2013)
+			// error if it is killed by the monitoring goroutine.
+			if _, err := dbClient.ExecuteFetch(action.Task, -1); err != nil {
+				failedAlterErr := err
+				// It's possible that we previously executed the ALTER but
+				// the subsequent DELETE of the post_copy_action record failed.
+				// For example, the context could be cancelled in-between.
+				// It's also possible that the user has modified the schema on
+				// the target side.
+				// If we get a duplicate key/index error then let's see if the
+				// index definitions that we would have added already exist in
+				// the table schema and if so move forward and delete the
+				// post_copy_action record.
+				if sqlErr, ok := err.(*mysql.SQLError); ok && sqlErr.Number() == mysql.ERDupKeyName {
+					stmt, err := sqlparser.ParseStrictDDL(action.Task)
+					if err != nil {
+						return failedAlterErr
+					}
+					alterStmt, ok := stmt.(*sqlparser.AlterTable)
+					if !ok {
+						return failedAlterErr
+					}
+					currentSKs, err := vr.getTableSecondaryKeys(ctx, tableName)
+					if err != nil {
+						return failedAlterErr
+					}
+					if len(currentSKs) < len(alterStmt.AlterOptions) {
+						return failedAlterErr
+					}
+					for _, alterOption := range alterStmt.AlterOptions {
+						addKey, ok := alterOption.(*sqlparser.AddIndexDefinition)
+						if !ok {
+							return failedAlterErr
+						}
+						found := false
+						for _, currentSK := range currentSKs {
+							if reflect.DeepEqual(addKey.IndexDefinition, currentSK) {
```

Comment:
Is it possible to use `sqlparser.Equals.RefOfIndexDefinition()` instead of `DeepQual`?

Reply:
Oooh, definitely. Didn't realize that existed. ❤️